system-tests/lib/benchspy/markdown_reporter.go

```diff
@@ -0,0 +1,413 @@
+}
+
+func createMultiSeriesBarChart(
+	title string,
+	dates []time.Time,
+	metricNames []string,
+	metricValues [][]float64,
+	maxValues []float64,
+	unitLabels []string,
+	scaleFactors []float64,
+) string {
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("\n## %s\n\n", title))
+	b.WriteString("| Date |")
+
+	// Add header row with metric names
+	for _, name := range metricNames {
+		b.WriteString(fmt.Sprintf(" %s |", name))
+	}
+	b.WriteString("\n")
+
+	// Add separator row
+	b.WriteString("| --- |")
+	for range metricNames {
+		b.WriteString(" --- |")
+	}
+	b.WriteString("\n")
+
+	// Add data rows
+	for i := 0; i < len(dates); i++ {
+		b.WriteString(fmt.Sprintf("| %s |", dates[i].Format("2006-01-02 15:04:05")))
+
+		for j := range metricNames {
+			if i < len(metricValues[j]) {
+				value := metricValues[j][i]
+				if j < len(scaleFactors) && scaleFactors[j] > 0 {
+					value /= scaleFactors[j]
+				}
+				unitLabel := ""
+				if j < len(unitLabels) {
+					unitLabel = unitLabels[j]
+				}
+				b.WriteString(fmt.Sprintf(" %s %.3f%s |",
+					bar(metricValues[j][i], maxValues[j], 10, "█"),
+					value,
+					unitLabel))
+			} else {
+				b.WriteString(" N/A |")
+			}
+		}
+		b.WriteString("\n")
+	}
+
+	return b.String()
+}
+
+// GenerateMarkdownReport generates a markdown report from benchmark data with configurable metric grouping
+func GenerateMarkdownReport(folder string, outputPath string, metricConfigs map[string]MetricConfig) error {
+	runs, err := readBenchmarkFiles(folder)
+	if err != nil {
+		return fmt.Errorf("error reading benchmark files: %w", err)
+	}
+
+	runMetrics, metricMaxValues := extractMetrics(runs)
+	content := generateReportContent(runs, runMetrics, metricMaxValues, metricConfigs)
+
+	// Write to .md file
+	if err := os.WriteFile(outputPath, []byte(content), 0644); err != nil {
+		return fmt.Errorf("error writing markdown file: %w", err)
+	}
+
+	return nil
+}
+
+// extractMetrics processes benchmark runs to extract metrics
+func extractMetrics(runs []report) ([]map[string]float64, map[string]float64) {
+	runMetrics := make([]map[string]float64, len(runs))
+
+	// Extract metrics from each run
+	for i, run := range runs {
+		runMetrics[i] = make(map[string]float64)
+		// For every query executor
+		for _, query := range run.QueryExecutors {
+			// For every metric in the query results
+			for metricName, metricValue := range query.QueryResults {
+				if query.Kind == "direct" {
+					processDirectMetric(runMetrics[i], metricName, metricValue)
+				} else if query.Kind == "prometheus" {
+					processPrometheusMetric(runMetrics[i], metricName, metricValue)
+				}
+			}
+		}
+	}
+
+	// Get max of every runMetric
+	metricMaxValues := findMaxMetricValues(runMetrics)
+
+	return runMetrics, metricMaxValues
+}
+
+// processDirectMetric processes a direct query metric
+func processDirectMetric(metrics map[string]float64, metricName string, metricValue interface{}) {
+	if strValue, ok := metricValue.(string); ok {
```

Comment:
can't we import the types that `BenchSpy` uses to unmarshall these values to?

Reply:
I don't believe that they are exported but we can do that in a follow up PR